content/blog/2025-05-13-bril-concurrency-extension.md

```diff
@@ -0,0 +1,163 @@
+    
+_In summary_, our RPC-based “shared-heap” prototype cleanly demonstrates the mechanics of true concurrency in Bril, but the messaging overhead severely limits speedups for fine-grained loops. Future work on shared‐memory backing or batched RPC will be necessary to turn this into a truly performant parallel interpreter.
+
+### 5.3 Escape Analysis Statistics
+
+```
+[escape] main: 1/1 allocs are thread-local   # solo allocation stays local
+[escape] TOTAL: 1/1 (100%) thread-local
+
+[escape] main: 0/1 allocs are thread-local   # spawn causes escape
+[escape] mixed: 1/2 allocs are thread-local  # one local, one escaping
+```
+
+These results confirm our pass correctly identifies thread-local allocations.
+
+6\. Challenges and Lessons Learned
+----------------------------------
+
+The switch from synchronous stubs to Web Workers introduced complexity in sharing the heap. Since workers cannot share JS memory directly, we built an RPC-based HeapProxy that marshals each alloc/load/store call. Debugging message ordering, promise lifetimes, and termination conditions consumed significant effort.
+I think my biggest learning was that I expected concurrency to speed everything up, but the 10 M-element test exposed how RPC overhead can negate parallel gains. 
+I also took some short cuts. A fully context-sensitive, interprocedural pass would be more precise, but the current intraprocedural, flow-insensitive approach seemed to be simplest useful thing I could do.
+
+7\. Future Work
+---------------
+
+Building on this foundation, several directions remain:
+
+*   **SharedArrayBuffer Heap**: Replace RPC with a SharedArrayBuffer–backed heap to eliminate message-passing overhead and unlock true multi-core speed-ups for large workloads.
+    
+*   **Par-for Loop Transform**: Implement a compiler pass that recognizes map/reduce patterns and automatically rewrites for loops into multiple spawn calls, guided by data-dependency analysis.
+    
+*   **Interprocedural Escape Analysis**: Extend escape.py to track pointers across function boundaries and calls, increasing precision and enabling stack-based allocation for truly local objects.
+    
+*   **Robust Testing Harness**: Integrate with continuous integration (CI) to run our concurrency and escape-analysis suites on every commit, ensuring regressions are caught early.
```

Comment:
Hmm, aren’t your interpreter tests enabled by default already? It also seems pretty easy to add your escape analysis tests to the main test suite, assuming they already use Turnt.